File: ts/src/assumptions.ts
import { fastSpread } from './fastSpread';

/**
 * Assumption validation framework for Pragmastat estimators.
 *
 * Assumption IDs (canonical priority order):
 *   1. validity - non-empty input with finite defined real values
 *   2. domain - parameter is outside its valid domain
 *   3. positivity - values must be strictly positive
 *   4. sparity - sample must be non tie-dominant (Spread > 0)
 *
 * When multiple assumptions are violated, the violation with highest priority
 * is reported. For two-sample functions, subject X is checked before Y.
 */

export const AssumptionId = {
  VALIDITY: 'validity',
  DOMAIN: 'domain',
  POSITIVITY: 'positivity',
  SPARITY: 'sparity',
} as const;

export type AssumptionId = (typeof AssumptionId)[keyof typeof AssumptionId];

export type Subject = 'x' | 'y' | 'misrate';

export interface Violation {
  id: AssumptionId;
  subject: Subject;
}

export class AssumptionError extends Error {
  readonly violation: Violation;

  constructor(violation: Violation) {
    const violationStr = `${violation.id}(${violation.subject})`;
    super(violationStr);
    this.name = 'AssumptionError';
    this.violation = violation;
  }

  static validity(subject: Subject): AssumptionError {
    return new AssumptionError({ id: AssumptionId.VALIDITY, subject });
  }

  static positivity(subject: Subject): AssumptionError {
    return new AssumptionError({ id: AssumptionId.POSITIVITY, subject });
  }

  static sparity(subject: Subject): AssumptionError {
    return new AssumptionError({ id: AssumptionId.SPARITY, subject });
  }

  static domain(subject: Subject): AssumptionError {
    return new AssumptionError({ id: AssumptionId.DOMAIN, subject });
  }
}

export function checkValidity(values: number[], subject: Subject): void {
  if (values.length === 0) {
    throw AssumptionError.validity(subject);
  }
  if (!values.every((v) => Number.isFinite(v))) {
    throw AssumptionError.validity(subject);
  }
}

export function checkPositivity(values: number[], subject: Subject): void {
  if (values.some((v) => v <= 0)) {
    throw AssumptionError.positivity(subject);
  }
}

export function checkSparity(values: number[], subject: Subject): void {
  if (values.length < 2) {
    throw AssumptionError.sparity(subject);
  }
  const spread = fastSpread(values);
  if (spread <= 0) {
    throw AssumptionError.sparity(subject);
  }
}

/**
 * Log-transforms an array. Throws AssumptionError if any value is non-positive.
 */
export function log(values: number[], subject: Subject): number[] {
  const result = new Array<number>(values.length);
  for (let i = 0; i < values.length; i++) {
    if (values[i] <= 0) {
      throw AssumptionError.positivity(subject);
    }
    result[i] = Math.log(values[i]);
  }
  return result;
}

File: go/assumptions.go
// Package pragmastat provides robust statistical estimators for real-world data analysis.
package pragmastat

import (
	"fmt"
	"math"
)

// AssumptionID identifies an assumption in canonical priority order.
type AssumptionID string

const (
	Validity   AssumptionID = "validity"
	Domain     AssumptionID = "domain"
	Positivity AssumptionID = "positivity"
	Sparity    AssumptionID = "sparity"
)

// Subject identifies which sample caused the violation in two-sample functions.
type Subject string

const (
	SubjectX       Subject = "x"
	SubjectY       Subject = "y"
	SubjectMisrate Subject = "misrate"
)

// DefaultMisrate is the default misclassification rate for bounds estimators.
const DefaultMisrate = 1e-3

// Violation represents a specific assumption violation.
type Violation struct {
	ID      AssumptionID
	Subject Subject
}

func (v Violation) String() string {
	return fmt.Sprintf("%s(%s)", v.ID, v.Subject)
}

// AssumptionError is the error type for assumption violations.
type AssumptionError struct {
	Violation Violation
}

func (e *AssumptionError) Error() string {
	return e.Violation.String()
}

func NewValidityError(subject Subject) *AssumptionError {
	return &AssumptionError{Violation: Violation{ID: Validity, Subject: subject}}
}

func NewPositivityError(subject Subject) *AssumptionError {
	return &AssumptionError{Violation: Violation{ID: Positivity, Subject: subject}}
}

func NewSparityError(subject Subject) *AssumptionError {
	return &AssumptionError{Violation: Violation{ID: Sparity, Subject: subject}}
}

func NewDomainError(subject Subject) *AssumptionError {
	return &AssumptionError{Violation: Violation{ID: Domain, Subject: subject}}
}

// Log log-transforms a slice. Returns error if any value is non-positive.
func Log[T Number](values []T, subject Subject) ([]float64, error) {
	result := make([]float64, len(values))
	for i, v := range values {
		fv := float64(v)
		if fv <= 0 {
			return nil, NewPositivityError(subject)
		}
		result[i] = math.Log(fv)
	}
	return result, nil
}
